### PR TITLE
[FLINK-34121][core] Introduce pipeline.force-kryo-avro to control whe…

### PR DESCRIPTION
**Pull request by reswqa**
…ther to force registration of Avro serializer with Kryo

## What is the purpose of the change

*Currently the avro serializers is registered with Kryo if flink-avro is on the classpath. This also happens if avro isn't even used by the job, be it due to a mistake in the dependencies setup, branching or flink-avro being in lib. This forces users to always provide flink-avro going forward for affected jobs, because on recovery Flink complains if any of the Kryo serializers cant be loaded.*

## Brief change log

  - *Introduce a new config option that controls whether or not register `avro` to `KryoSerializer`.*

## Verifying this change

Basically covered by existing tests, while also adding some unit tests.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: yes
  - The serializers: yes
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no

### CODE REVIEW

File: flink-core/src/main/java/org/apache/flink/api/common/ExecutionConfig.java [hunk 1]
```diff
@@ -666,6 +666,18 @@ public boolean isForceKryoEnabled() {
...
+    }
+
+    private void setForceAvroKryo(boolean forceAvroKryo) {
+        configuration.set(PipelineOptions.FORCE_KRYO_AVRO, forceAvroKryo);
+    }
+
+    public Optional<Boolean> isForceAvroKryoEnable() {
+        return configuration.getOptional(PipelineOptions.FORCE_KRYO_AVRO);
+    }
+
```
Comment (JunRuiLee) [inline flink-core/src/main/java/org/apache/flink/api/common/ExecutionConfig.java:680]: I am curious why there are three public methods introduced that were not included in FLIP-398.

File: docs/layouts/shortcodes/generated/pipeline_configuration.html [hunk 2]
```diff
@@ -56,6 +56,12 @@
             <td>Boolean</td>
             <td>If enabled, forces TypeExtractor to use Kryo serializer for POJOS even though we could analyze as POJO. In some cases this might be preferable. For example, when using interfaces with subclasses that cannot be analyzed as POJO.</td>
         </tr>
+        <tr>
+            <td><h5>pipeline.force-kryo-avro</h5></td>
+            <td style="word-wrap: break-word;">(none)</td>
+            <td>Boolean</td>
+            <td>Force register avro classes in kryo serializer. <br /><br />Important: Make sure to include the flink-avro module. Otherwise, nothing will be registered. For backward compatibility, the default value is empty to conform to the behavior of the older version. That is, always register avro with kryo, and if flink-avro is not in the class path, register a dummy Serializer. In flink-2.0, we will set the default value to true</td>
```
Comment (X-czh) [inline docs/layouts/shortcodes/generated/pipeline_configuration.html:63]:
```suggestion
            <td>Forces Flink to register avro classes in kryo serializer. <br /><br />Important: Make sure to include the flink-avro module. Otherwise, nothing will be registered. For backward compatibility, the default value is empty to conform to the behavior of the older version. That is, always register avro with kryo, and if flink-avro is not in the class path, register a dummy serializer. In Flink 2.0, we will set the default value to true.</td>
```

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/catalog/DataTypeFactoryImpl.java [hunk 3]
```diff
@@ -141,6 +141,16 @@ private static Supplier<ExecutionConfig> createSerializerExecutionConfig(
...
 
+                executionConfig
+                        .getSerializerConfig()
+                        .isForceAvroKryo()
+                        .filter(isForceAvroKryo -> isForceAvroKryo)
+                        .ifPresent(
+                                ignore ->
+                                        newExecutionConfig
+                                                .getSerializerConfig()
+                                                .setForceAvroKryo(true));
```
Comment (X-czh) [inline flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/catalog/DataTypeFactoryImpl.java:152]: What if `forceAvroKryo` is explicitly set to false?

**Comment by X-czh** [comment]
Thanks for the contribution. LGTM

**Review by X-czh — APPROVED** [review]
